File: typescript/src/webhooks.ts
import { createHmac, timingSafeEqual } from "node:crypto";

export type WebhookSecret = string | Uint8Array;
export type WebhookHeader = string | readonly string[];

export interface WebhookVerificationOptions {
  timestampHeader?: string;
  secrets: WebhookSecret | readonly WebhookSecret[];
  toleranceSeconds?: number;
  now?: number;
  allowLegacy?: boolean;
}

export class WebhookVerificationError extends Error {
  constructor(message = "invalid or stale DexCost webhook signature") {
    super(message);
    this.name = "WebhookVerificationError";
  }
}

const MAX_SECRETS = 8;
const MAX_SIGNATURES = 16;

function payloadBytes(payload: unknown): Uint8Array {
  if (!(payload instanceof Uint8Array)) {
    throw new TypeError("webhook payload must be the unmodified raw bytes");
  }
  return payload;
}

function secretBytes(value: WebhookVerificationOptions["secrets"]): Uint8Array[] {
  const raw = typeof value === "string" || value instanceof Uint8Array ? [value] : value;
  if (!Array.isArray(raw) || raw.length < 1 || raw.length > MAX_SECRETS) {
    throw new TypeError(`webhook verification supports 1 to ${MAX_SECRETS} secrets`);
  }
  return raw.map((secret) => {
    const encoded = typeof secret === "string" ? Buffer.from(secret, "utf8") : secret;
    if (!(encoded instanceof Uint8Array) || encoded.byteLength < 1 || encoded.byteLength > 1024) {
      throw new TypeError("each webhook secret must contain 1 to 1024 bytes");
    }
    return encoded;
  });
}

function headerSignatures(header: WebhookHeader, versioned: boolean): Buffer[] {
  const values = typeof header === "string" ? [header] : header;
  if (!Array.isArray(values)) throw new TypeError("webhook signature header must be a string or sequence");
  const entries = values.flatMap((value) => {
    if (typeof value !== "string") throw new TypeError("webhook signature header values must be strings");
    return value.split(",").map((part) => part.trim()).filter(Boolean);
  });
  if (entries.length > MAX_SIGNATURES) {
    throw new TypeError(`webhook verification supports at most ${MAX_SIGNATURES} signatures`);
  }
  return entries.flatMap((entry) => {
    let digest = entry;
    const separator = entry.indexOf("=");
    if (separator >= 0) {
      const scheme = entry.slice(0, separator).toLowerCase();
      digest = entry.slice(separator + 1);
      if (scheme !== (versioned ? "v1" : "sha256")) return [];
    } else if (versioned) return [];
    if (!/^[0-9a-fA-F]{64}$/.test(digest)) return [];
    return [Buffer.from(digest, "hex")];
  });
}

function timestamp(value: string, toleranceSeconds: number, now?: number): string {
  if (!/^[0-9]+$/.test(value)) throw new TypeError("webhook timestamp must be unix seconds");
  if (!Number.isFinite(toleranceSeconds) || toleranceSeconds < 0 || toleranceSeconds > 86_400) {
    throw new RangeError("webhook toleranceSeconds must be between 0 and 86400");
  }
  const current = now ?? Date.now() / 1000;
  if (!Number.isFinite(current)) throw new TypeError("webhook now must be unix seconds");
  const signedAt = Number(value);
  if (!Number.isSafeInteger(signedAt) || Math.abs(current - signedAt) > toleranceSeconds) {
    throw new RangeError("webhook timestamp is outside the accepted tolerance");
  }
  return value;
}

export function verifyWebhookSignature(
  payload: Uint8Array,
  signatureHeader: WebhookHeader,
  options: WebhookVerificationOptions,
): boolean {
  try {
    const bytes = payloadBytes(payload);
    const secrets = secretBytes(options.secrets);
    const versioned = options.timestampHeader !== undefined;
    if (!versioned && options.allowLegacy !== true) return false;
    const signatures = headerSignatures(signatureHeader, versioned);
    if (signatures.length === 0) return false;
    const message = versioned
      ? Buffer.concat([
        Buffer.from(timestamp(options.timestampHeader!, options.toleranceSeconds ?? 300, options.now), "ascii"),
        Buffer.from("."), Buffer.from(bytes),
      ])
      : Buffer.from(bytes);
    for (const secret of secrets) {
      const expected = createHmac("sha256", secret).update(message).digest();
      for (const received of signatures) {
        if (received.length === expected.length && timingSafeEqual(expected, received)) return true;
      }
    }
    return false;
  } catch {
    return false;
  }
}

export function assertWebhookSignature(
  payload: Uint8Array,
  signatureHeader: WebhookHeader,
  options: WebhookVerificationOptions,
): void {
  if (!verifyWebhookSignature(payload, signatureHeader, options)) {
    throw new WebhookVerificationError();
  }
}

File: python/src/dexcost/webhooks.py
"""Timing-safe DexCost webhook verification with replay protection."""

from __future__ import annotations

import hashlib
import hmac
import time
from collections.abc import Sequence
from typing import TypeAlias, cast

WebhookSecret: TypeAlias = str | bytes
WebhookHeader: TypeAlias = str | Sequence[str]

_MAX_SECRETS = 8
_MAX_SIGNATURES = 16


class WebhookVerificationError(ValueError):
    """A webhook did not satisfy the signature and freshness contract."""


def _payload_bytes(payload: bytes | bytearray | memoryview) -> bytes:
    if not isinstance(payload, (bytes, bytearray, memoryview)):
        raise TypeError("webhook payload must be the unmodified raw bytes")
    return bytes(payload)


def _secret_bytes(secrets: WebhookSecret | Sequence[WebhookSecret]) -> tuple[bytes, ...]:
    raw_secrets: Sequence[WebhookSecret]
    if isinstance(secrets, (str, bytes)):
        raw_secrets = (secrets,)
    elif isinstance(secrets, Sequence):
        raw_secrets = cast("Sequence[WebhookSecret]", secrets)
    else:
        raise TypeError("webhook secrets must be a secret or sequence of secrets")
    if not 1 <= len(raw_secrets) <= _MAX_SECRETS:
        raise ValueError(f"webhook verification supports 1 to {_MAX_SECRETS} secrets")
    encoded: list[bytes] = []
    for secret in raw_secrets:
        value = secret.encode("utf-8") if isinstance(secret, str) else secret
        if not isinstance(value, bytes) or not 1 <= len(value) <= 1024:
            raise ValueError("each webhook secret must contain 1 to 1024 bytes")
        encoded.append(value)
    return tuple(encoded)


def _header_entries(header: WebhookHeader) -> tuple[str, ...]:
    values = (header,) if isinstance(header, str) else header
    if not isinstance(values, Sequence):
        raise TypeError("webhook signature header must be a string or sequence")
    entries: list[str] = []
    for value in values:
        if not isinstance(value, str):
            raise TypeError("webhook signature header values must be strings")
        entries.extend(part.strip() for part in value.split(",") if part.strip())
        if len(entries) > _MAX_SIGNATURES:
            raise ValueError(f"webhook verification supports at most {_MAX_SIGNATURES} signatures")
    return tuple(entries)


def _hex_signatures(header: WebhookHeader, *, versioned: bool) -> tuple[str, ...]:
    signatures: list[str] = []
    for entry in _header_entries(header):
        if "=" in entry:
            scheme, digest = entry.split("=", 1)
            accepted = scheme.lower() == ("v1" if versioned else "sha256")
            if not accepted:
                continue
        elif versioned:
            continue
        else:
            digest = entry
        normalized = digest.lower()
        if len(normalized) == 64 and all(
            character in "0123456789abcdef" for character in normalized
        ):
            signatures.append(normalized)
    return tuple(signatures)


def _timestamp(value: str, tolerance_seconds: float, now: float | None) -> str:
    if not isinstance(value, str) or not value or not value.isascii() or not value.isdigit():
        raise ValueError("webhook timestamp must be unix seconds")
    if (
        isinstance(tolerance_seconds, bool)
        or not isinstance(tolerance_seconds, (int, float))
        or not 0 <= tolerance_seconds <= 86_400
    ):
        raise ValueError("webhook tolerance_seconds must be between 0 and 86400")
    signed_at = int(value)
    current = time.time() if now is None else now
    if not isinstance(current, (int, float)) or isinstance(current, bool):
        raise TypeError("webhook now must be unix seconds")
    if abs(float(current) - signed_at) > float(tolerance_seconds):
        raise ValueError("webhook timestamp is outside the accepted tolerance")
    return value


def _matches(message: bytes, signatures: tuple[str, ...], secrets: tuple[bytes, ...]) -> bool:
    if not signatures:
        return False
    for secret in secrets:
        expected = hmac.new(secret, message, hashlib.sha256).hexdigest()
        for received in signatures:
            if hmac.compare_digest(expected, received):
                return True
    return False


def verify_webhook_signature(
    payload: bytes | bytearray | memoryview,
    signature_header: WebhookHeader,
    *,
    timestamp_header: str | None,
    secrets: WebhookSecret | Sequence[WebhookSecret],
    tolerance_seconds: float = 300,
    now: float | None = None,
    allow_legacy: bool = False,
) -> bool:
    """Return whether a raw payload has a fresh valid DexCost signature.

    New deliveries use ``X-Dexcost-Signature-V1`` plus
    ``X-Dexcost-Webhook-Timestamp``. Body-only legacy signatures are accepted
    only when ``allow_legacy=True`` because they cannot prevent replay.
    Malformed or adversarial input returns ``False`` and never raises.
    """
    try:
        raw_payload = _payload_bytes(payload)
        resolved_secrets = _secret_bytes(secrets)
        if timestamp_header is not None:
            timestamp = _timestamp(timestamp_header, tolerance_seconds, now)
            signatures = _hex_signatures(signature_header, versioned=True)
            return _matches(
                timestamp.encode("ascii") + b"." + raw_payload,
                signatures,
                resolved_secrets,
            )
        if not allow_legacy:
            return False
        signatures = _hex_signatures(signature_header, versioned=False)
        return _matches(raw_payload, signatures, resolved_secrets)
    except (TypeError, ValueError, OverflowError):
        return False


def assert_webhook_signature(
    payload: bytes | bytearray | memoryview,
    signature_header: WebhookHeader,
    *,
    timestamp_header: str | None,
    secrets: WebhookSecret | Sequence[WebhookSecret],
    tolerance_seconds: float = 300,
    now: float | None = None,
    allow_legacy: bool = False,
) -> None:
    """Raise one non-oracular error when webhook verification fails."""
    if not verify_webhook_signature(
        payload,
        signature_header,
        timestamp_header=timestamp_header,
        secrets=secrets,
        tolerance_seconds=tolerance_seconds,
        now=now,
        allow_legacy=allow_legacy,
    ):
        raise WebhookVerificationError("invalid or stale DexCost webhook signature")


__all__ = [
    "WebhookHeader",
    "WebhookSecret",
    "WebhookVerificationError",
    "assert_webhook_signature",
    "verify_webhook_signature",
]
